[#3946] Support conversion of HttpMessage and HttpContent to HTTP/2 Frames

blucas opened the pull request:
Motivation:

```HttpToHttp2ConnectionHandler``` only converts ```FullHttpMessage``` to HTTP/2 Frames. This does not support other use cases such as adding a ```HttpContentCompressor``` to the pipeline, which writes ```HttpMessage``` and ```HttpContent```.

Additionally ```HttpToHttp2ConnectionHandler``` ignores converting and sending HTTP trailing headers, which is a bug as the HTTP/2 spec states that they should be sent.

Modifications:

Update ```HttpToHttp2ConnectionHandler``` to support converting ```HttpMessage``` and ```HttpContent``` to HTTP/2 Frames.
Additionally, include an extra call to ```writeHeaders``` if the message includes trailing headers.

Result:

One can now write ```HttpMessage``` and ```HttpContent``` (http chunking) down the pipeline and they will be converted to HTTP/2 Frames.  If any trailing headers exist, they will be converted and sent as well.

netkins commented:
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

blucas commented:
@nmittler @Scottmitch please take a look.

normanmaurer commented:
@netkins acept

normanmaurer commented:
@netkins accept

blucas commented:
I don't know why the tests are failing... They pass on my local

blucas commented:
@nmittler @Scottmitch ping

nmittler commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+
+                    if (isEmpty) {
+                        encoder.writeData(ctx, currentStreamId, httpContent.content(), 0, true, promise);
+                        release = false;
```
Should probably set `release = false` just before the write.

normanmaurer replied:
I think after is good so as writeData() may throw an exception before write any data.

Scottmitch replied:
+1 to @nmittler. the writeData is responsible for releasing the data (even in the presence of exceptions) per the interface contract.  We should avoid double releasing here and elsewhere in this PR.

Scottmitch replied:
I think @nmittler's suggestion (https://github.com/netty/netty/pull/3975#discussion_r34686595) to consolidate the code will help ensure we are releasing correctly.

nmittler commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
-            boolean hasData = httpMsg.content().isReadable();
-            boolean httpMsgNeedRelease = true;
+
+        if (msg instanceof HttpMessage) {
```
Just to aid in readability, can you separate these blocks out into separate methods (e.g. `writeHttpMessage` and `writeHttpContent`)?

normanmaurer replied:
+1

nmittler commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+
+                        promiseAggregator.doneAllocatingPromises();
+                    } else {
+                        encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, true, promise);
```
I think we should be able to simplify the logic a bit.  Could we do something like this:

```java
HttpContent content = null;
// Just always use the promise aggregator.
SimpleChannelPromiseAggregator promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel, ctx.executor);

if (msg instanceof HttpMessage) {
    HttpMessage httpMsg = (HttpMessage) msg;

    // Write headers using the promise aggregator.
}

if (msg instanceof HttpContent) {
    HttpContent httpContent = (HttpContent) msg;

    // Write the data.

    if (msg instanceof LastHttpContent) {
        LastHttpContent lastHttpContent = (LastHttpContent) msg;

        // Write Data

        // Write trailers if provided.
    }
}

promiseAggregator.doneAllocatingPromises();
```

WDYT?

normanmaurer replied:
yeah maybe not a bad idea

Scottmitch replied:
+1. This way we only have to worry about releasing in the `HttpContent` block where you would expect it.

nmittler commented:
@blucas Just took a first pass.  Looks pretty good overall!

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+                    FullHttpMessage fullHttpMsg = (FullHttpMessage) httpMsg;
+                    boolean hasData = fullHttpMsg.content().isReadable();
+                    Http2Headers trailingHeaders = fullHttpMsg.trailingHeaders().isEmpty() ? null :
+                            HttpUtil.toHttp2Headers((LastHttpContent) fullHttpMsg);
```
I think the cast is not needed

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java:
```diff
@@ -324,6 +326,32 @@ public boolean visit(Entry<CharSequence, CharSequence> entry) throws Exception {
+        inHeaders.forEachEntry(new EntryVisitor() {
+            @Override
+            public boolean visit(Entry<CharSequence, CharSequence> entry) throws Exception {
+                final AsciiString aName = AsciiString.of(entry.getKey()).toLowerCase();
```
toLowerCase(Locale.US) ?

Scottmitch replied:
I'm not sure this is necessary. It is converted to an `AsciiString` which implies a certain encoding already. WDYT?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java:
```diff
@@ -324,6 +326,32 @@ public boolean visit(Entry<CharSequence, CharSequence> entry) throws Exception {
     /**
+     * Converts HTTP/1.1 trailing headers into HTTP/2 headers
+     */
+    public static Http2Headers toHttp2Headers(LastHttpContent in) throws Exception {
```
Change signature to accept HttpHeaders ?

Scottmitch replied:
+1. This code should also be shared with the other method `toHttp2Headers` that also has this code.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+            boolean release = true;
+            SimpleChannelPromiseAggregator promiseAggregator = null;
+            try {
+                if (currentStreamId == 0) {
```
I'm not sure we should check this here (or worry about resetting `currentStreamId = 0`).  The HTTP/2 codec is responsible for enforcing this.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+                    }
+
+                    // Reset the stream id
+                    currentStreamId = 0;
```
Lets remove this per https://github.com/netty/netty/pull/3975#discussion-diff-34731790

Scottmitch commented:
@blucas - Thanks for the PR!  Finished first pass...looks good so far. Ping me when rdy for another review.

blucas commented:
@nmittler @Scottmitch @normanmaurer addressed your comments please have another look.

I have re-written the method as per Nathan's suggestion. There is a lot less going on now, so may make more sense.

nmittler commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+
+            promiseAggregator.doneAllocatingPromises();
+        } catch (Throwable t) {
+            promiseAggregator.setFailure(t);
```
Need to call `promiseAggregator.donAllocatingPromises()` here too.

Scottmitch replied:
@nmittler - I don't think we should have to call `donAllocatingPromises()` in the event of failure. I think the `SimpleChannelPromiseAggregator` should support termination by calling `setFailure(..)`.

nmittler replied:
Ah ok ... wasn't entirely sure.  Haven't looked at that class in a while :)

nmittler commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+                if (!endStream && msg instanceof LastHttpContent) {
+                    // Convert and write any trailing headers.
+                    final LastHttpContent lastContent = (LastHttpContent) httpContent;
+                    final Http2Headers headers = lastContent.trailingHeaders().isEmpty() ? null :
```
Maybe further simplify the logic here to something like:

```java
if (msg instanceof HttpMessage) {
    // Write headers.
}

if (!endStream && msg instanceof HttpContent) {
    // If this is the last content, extract the trailers.
    boolean lastContent = false;
    Http2Headers trailers = null;
    if (msg instanceof LastHttpContent) {
        lastContent = true;
        trailers = HttpUtil.toHttp2Headers(lastContent.trailingHeaders());
        // TODO: check to see if trailers can really be null here.
        if (trailers == null || trailers.isEmpty()) {
            trailers = null;
        }
    }

    release = false;
    encoder.writeData(..., trailers != null, ...);
    if (trailers != null) {
        encoder.writeHeaders(..., true, ...);
    }

    endStream = lastContent;
}
```

Scottmitch replied:
+1. I think we can drop the `null` check for `trailers` too.

Scottmitch commented:
@blucas - Looks like there is also a unit test failure on the build.

blucas commented:
@nmittler @Scottmitch PTAL

I have simplified the code further as per Nathan's suggestion.

I hope I've fixed the unit test failure.  The tests have always passed on my machine, so I can only think it is a 'timing' issue.  I added a ```CountDownLatch``` for trailers, and increased the request countdown to 4 in my tests.

I have tried to test the trailers in both Firefox (38.0.5) and Chrome (Version 43.0.2357.134 (64-bit)). Firefox works as expected (the page is rendered properly).  Unfortunately it looks like Chrome does not expect to receive a ```HEADERS``` frame after a ```DATA``` frame.  I think this is a bug with Chrome.  Search for "Additional headers received for request/response stream" in the HTTP/2 Frame output below:

```
7167: HTTP2_SESSION
www.mysite.com:443 (DIRECT)
Start Time: 2015-07-18 12:17:32.052

t=888877 [st=    0] +HTTP2_SESSION  [dt=30104]
                     --> host = "www.mysite.com:443"
                     --> proxy = "DIRECT"
t=888877 [st=    0]    HTTP2_SESSION_INITIALIZED
                       --> protocol = "h2"
                       --> source_dependency = 7166 (SOCKET)
t=888877 [st=    0]    HTTP2_SESSION_SEND_SETTINGS
                       --> settings = ["[id:3 flags:0 value:1000]","[id:4 flags:0 value:10485760]"]
t=888877 [st=    0]    HTTP2_STREAM_UPDATE_RECV_WINDOW
                       --> delta = 10420225
                       --> window_size = 10485760
t=888877 [st=    0]    HTTP2_SESSION_SENT_WINDOW_UPDATE_FRAME
                       --> delta = 10420225
                       --> stream_id = 0
t=888878 [st=    1]    HTTP2_SESSION_SEND_HEADERS
                       --> fin = true
                       --> :authority: www.mysite.com
                           :method: GET
                           :path: /
                           :scheme: https
                           accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
                           accept-encoding: gzip, deflate, sdch
                           accept-language: en-US,en;q=0.8
                           cache-control: max-age=0
                           user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36
                       --> priority = 0
                       --> stream_id = 1
                       --> unidirectional = false
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTINGS
                       --> clear_persisted = false
                       --> host = "www.mysite.com:443"
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTING
                       --> flags = 0
                       --> id = 1
                       --> value = 4096
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTING
                       --> flags = 0
                       --> id = 3
                       --> value = 2147483647
t=888881 [st=    4]    HTTP2_SESSION_UPDATE_STREAMS_SEND_WINDOW_SIZE
                       --> delta_window_size = 0
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTING
                       --> flags = 0
                       --> id = 4
                       --> value = 65535
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTING
                       --> flags = 0
                       --> id = 5
                       --> value = 16384
t=888881 [st=    4]    HTTP2_SESSION_RECV_SETTING
                       --> flags = 0
                       --> id = 6
                       --> value = 2147483647
t=888974 [st=   97]    HTTP2_SESSION_RECV_HEADERS
                       --> fin = false
                       --> :status: 403
                           content-length: 7
                           test: denied=YesMozilla/5.0 (Macintosh; Intel Mac OS X 10_7_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36
                           trailer: testing
                       --> stream_id = 1
t=888975 [st=   98]    HTTP2_SESSION_RECV_DATA
                       --> fin = false
                       --> size = 7
                       --> stream_id = 1
t=888975 [st=   98]    HTTP2_SESSION_UPDATE_RECV_WINDOW
                       --> delta = -7
                       --> window_size = 10485753
t=888975 [st=   98]    HTTP2_SESSION_RECV_HEADERS
                       --> fin = true
                       --> testing: 123
                       --> stream_id = 1
t=888975 [st=   98]    HTTP2_SESSION_SEND_RST_STREAM
                       --> description = "Additional headers received for request/response stream"
                       --> status = 1
                       --> stream_id = 1
t=888975 [st=   98]    HTTP2_SESSION_RECV_DATA
                       --> fin = true
                       --> size = 0
                       --> stream_id = 1
t=888975 [st=   98]    HTTP2_STREAM_UPDATE_RECV_WINDOW
                       --> delta = 7
                       --> window_size = 10485760
t=918981 [st=30104]    HTTP2_SESSION_GOAWAY
                       --> active_streams = 0
                       --> last_accepted_stream_id = 1
                       --> status = 0
                       --> unclaimed_streams = 0
t=918981 [st=30104]    HTTP2_SESSION_CLOSE
                       --> description = "Finished going away"
                       --> net_error = 0 (?)
t=918981 [st=30104]    HTTP2_SESSION_POOL_REMOVE_SESSION
                       --> source_dependency = 7167 (HTTP2_SESSION)
t=918981 [st=30104] -HTTP2_SESSION
```

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
-                    promise.tryFailure(t);
-                } else {
-                    promiseAggregator.setFailure(t);
+                endStream = msg instanceof FullHttpMessage && !((FullHttpMessage) msg).content().isReadable();
```
hmm... should this always end the stream if a FullHttpMessage is used ?

blucas replied:
As far as I'm aware, yes.  If the FullHttpMessage don't not contain any data, the only HTTP/2 frame to write is the HEADERS frame.  You can see the old code did the same thing.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java:
```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
-                    httpMsg.release();
+
+                // Write the data
+                release = false;
```
I would move the release directly before writeData(...) call. Just in case

blucas replied:
You are right.  I'll fix that.

nmittler commented:
@blucas LGTM 

@Scottmitch do you want to take another look?

normanmaurer commented:
@blucas looks good...  could you squash the commits ?

@Scottmitch please take a final look.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java:
```diff
@@ -304,6 +305,17 @@ public static Http2Headers toHttp2Headers(FullHttpMessage in) throws Exception {
         }
 
         // Add the HTTP headers which have not been consumed above
+        out.add(toHttp2Headers(inHeaders));
```
nit: just return here.

```java
return out.add(toHttp2Headers(inHeaders));
```

Scottmitch commented:
@blucas - Just 1 nit. Please squash your commits and we can cherry-pick.

blucas commented:
@normanmaurer @Scottmitch All commits squashed :)

normanmaurer commented:
@blucas cool... let me take care

normanmaurer commented:
@blucas... thanks!

Cherry-picked into master (3ec02bf7460c3fda4d382bdf4c3b9ef4c67704e1) and 4.1 (57d28dd42178b35194037a4ded7f26e20e676c62)
